be/src/exec/pipeline/scan/morsel.h

```diff
@@ -478,6 +492,31 @@ class LogicalSplitMorselQueue final : public SplitMorselQueue {
     ShortKeyIndexGroupIterator _next_lower_block_iter;
 };
 
+class DynamicMorselQueue final : public MorselQueue {
+public:
+    explicit DynamicMorselQueue(Morsels&& morsels) {
+        append_morsels(std::move(morsels));
+        _size = _num_morsels = _queue.size();
+    }
+    ~DynamicMorselQueue() override = default;
+    bool empty() const override { return _size == 0; }
```

Comment:
There may be race conditions.
- Thread#1 gets a morsel and ungets it.
- Before thread#1 ungets, thread#2 checks `empty()`, which returning false.

Reply:
since unget is only used in single thread, I guess that will not happen.
I'll rethink of it and refactor this code.